fix(operator): support private VPC S3 endpoints in Loki operator

puretension opened the pull request:
**What this PR does / why we need it**:

This PR fixes the Loki operator's overly strict S3 endpoint validation that was rejecting private VPC S3 endpoints. The operator was only accepting the standard AWS S3 endpoint format (`https://s3.region.amazonaws.com`) and failing to reconcile when users configured private VPC endpoints in OpenShift environments.

The fix updates the `validateS3Endpoint` function to also accept VPC endpoint formats:
- `https://bucket.vpce-*-region.s3.region.vpce.amazonaws.com` (bucket-specific VPC endpoint)
- `https://vpce-*-region.s3.region.vpce.amazonaws.com` (general VPC endpoint)

**Which issue(s) this PR fixes**:
Fixes #19243

**Special notes for your reviewer**:

This change maintains full backward compatibility with existing standard AWS S3 endpoints while adding support for VPC endpoints. The validation logic now:

1. First checks for standard AWS S3 endpoint format
2. Then checks for VPC endpoint patterns (contains `.vpce.amazonaws.com` and the specified region)
3. Only rejects endpoints that match neither pattern

The fix includes comprehensive test cases covering both VPC endpoint formats to ensure the validation works correctly.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added (not applicable - internal validation logic)
- [x] Tests updated (added comprehensive test cases for VPC endpoints)
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md` (not applicable - this is a bug fix that enables previously broken functionality)
- [x] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively (not applicable - no configuration changes)

## Changes Made

### Code Changes
- **Modified** `operator/internal/handlers/internal/storage/secrets.go`:
  - Updated `validateS3Endpoint()` function to support VPC endpoint validation
  - Added logic to detect VPC endpoint patterns
  - Maintained backward compatibility with standard AWS endpoints

### Test Changes  
- **Added** comprehensive test cases in `operator/internal/handlers/internal/storage/secrets_test.go`:
  - Test case for bucket-specific VPC endpoint format
  - Test case for general VPC endpoint format
  - Both test cases verify correct configuration parsing

## Validation Logic

**Before (rejected VPC endpoints):**
```go
validEndpoint := fmt.Sprintf("https://s3.%s%s", region, awsEndpointSuffix)
if endpoint != validEndpoint {
   return fmt.Errorf("%w: %s", errS3EndpointAWSInvalid, validEndpoint)
}
```

**After (accepts both standard and VPC endpoints):**
```go
// Check standard AWS S3 endpoint
validEndpoint := fmt.Sprintf("https://s3.%s%s", region, awsEndpointSuffix)
if endpoint == validEndpoint {
   return nil
}

// Check VPC endpoint format
if strings.Contains(endpoint, ".vpce.amazonaws.com") && strings.Contains(
endpoint, region) {
   return nil
}

// Reject if neither format matches
return fmt.Errorf("%w: %s", errS3EndpointAWSInvalid, validEndpoint)

```

## Impact

This fix enables users to successfully deploy Loki in OpenShift environments with private VPC S3 endpoints, resolving the operator reconciliation failures described in issue #19243.

puretension committed:
fix(operator): support private VPC S3 endpoints in Loki operator

This PR fixes the Loki operator's overly strict S3 endpoint validation that was rejecting private VPC S3 endpoints. The operator was only accepting the standard AWS S3 endpoint format and failing to reconcile when users configured private VPC endpoints in OpenShift environments.

The fix updates the validateS3Endpoint function to accept VPC endpoint formats while preventing bucket-specific VPC endpoints that could cause folder creation issues:
- Allows: https://vpce-*-region.s3.region.vpce.amazonaws.com (general VPC endpoint)
- Rejects: https://bucket.vpce-*-region.s3.region.vpce.amazonaws.com (bucket-specific VPC endpoint)

This maintains full backward compatibility with existing standard AWS S3 endpoints while adding support for VPC endpoints without the bucket name prefix that could lead to unwanted folder creation in S3 buckets.

Fixes #19243

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

JoaoBraveCoding reviewed (commented):
First of all thank you so much for contributing to the project! 🙏 It's much appreciated!

Overall look good just a small comment and I would like to ask you if you could rework the commit message/PR tittle so it shows `fix(operator): ...` so that your change is included in the release notes.

puretension committed:
fix(operator): improve VPC endpoint validation error flow

- Follow Go's 'Indent Error Flow' pattern by handling error cases first
- Provide more specific error message for bucket-specific VPC endpoints
- Maintain backward compatibility with existing validation logic
- Update test to reflect clearer error messaging

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
fix(operator): refactor S3 endpoint validation for better readability

- Use shorter variable names following Go conventions (u, host)
- Add clear English comments for each validation section
- Improve function signature with grouped parameters
- Maintain same validation logic and error handling

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
fix(operator): address code review feedback for VPC endpoint validation

- Revert variable name from 'u' to 'parsedURL' for better readability
- Keep AWS validation logic within conditional block structure
- Define VPC bucket name error as constant for consistency

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension commented on operator/internal/handlers/internal/storage/secrets.go:
```diff
@@ -482,36 +482,46 @@ func extractS3ConfigSecret(s *corev1.Secret, credentialMode lokiv1.CredentialMod
-	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
-		return fmt.Errorf("%w: %s", errS3EndpointUnsupportedScheme, parsedURL.Scheme)
+	// Non-AWS S3 compatible endpoints (e.g., MinIO) - no further validation needed
+	if !strings.HasSuffix(endpoint, awsEndpointSuffix) {
```
Makes sense! Restructured the AWS validation as suggested.

JoaoBraveCoding committed:
refactor to use regex + e2e testing

JoaoBraveCoding reviewed (commented):
Internally me and @xperimental discussed briefly your PR and he suggested trying out doing a validation with a regex, so based on your PR I did a commit and opened a PR https://github.com/puretension/loki/pull/1. While I was at it I also tested this against an actual VPC bucket to check if we were adding the right support for the endpoint. I think the only thing we might be missing is supporting `accesspoint` and `control` not only `bucket` but today I was only left with time to test `bucket`, took this info from here https://docs.aws.amazon.com/AmazonS3/latest/userguide/privatelink-interface-endpoints.html#accessing-bucket-and-aps-from-interface-endpoints.

puretension committed:
Merge pull request #1 from JoaoBraveCoding/vpc-buckets-review

refactor to use regex + e2e testing

JoaoBraveCoding commented on operator/internal/handlers/internal/storage/secrets.go:
```diff
@@ -42,7 +43,9 @@ var (
-	errS3EndpointAWSInvalid        = errors.New("endpoint for AWS S3 must include correct region")
+	errS3EndpointAWSNoRegion       = errors.New("endpoint for AWS S3 must include correct region")
+	errS3EndpointNoBucketName      = errors.New("bucket name must not be included in AWS S3 endpoint URL")
+	errS3EndpointAWSInvalid        = errors.New("endpoint for AWS S3 is invalid, must match either https://s3.region.amazonaws.com or https://vpce-id.s3.region.vpce.amazonaws.com")
```
Forgot to update this error message
```suggestion
	errS3EndpointAWSInvalid        = errors.New("endpoint for AWS S3 is invalid, must match either https://s3.region.amazonaws.com or https://bucket.vpce-id.s3.region.vpce.amazonaws.com")
```

JoaoBraveCoding commented on operator/internal/handlers/internal/storage/secrets_test.go:
```diff
@@ -629,7 +629,49 @@ func TestS3Extract(t *testing.T) {
 				},
 			},
-			wantError: "endpoint for AWS S3 must include correct region: https://s3.region.amazonaws.com",
+			wantError: "endpoint for AWS S3 is invalid, must match either https://s3.region.amazonaws.com or https://vpce-id.s3.region.vpce.amazonaws.com: http://region.amazonaws.com",
```
```suggestion
			wantError: "endpoint for AWS S3 is invalid, must match either https://s3.region.amazonaws.com or https://bucket.vpce-id.s3.region.vpce.amazonaws.com: http://region.amazonaws.com",
```

JoaoBraveCoding commented on operator/internal/handlers/internal/storage/secrets_test.go:
```diff
@@ -629,7 +629,49 @@ func TestS3Extract(t *testing.T) {
+			secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Data: map[string][]byte{
+					"endpoint":          []byte("https://vpce-1234567abc.s3.us-east-1.vpce.amazonaws.com"),
```
```suggestion
					"endpoint":          []byte("https://bucket.vpce-1234567abc.s3.us-east-1.vpce.amazonaws.com"),
```

JoaoBraveCoding commented on operator/internal/handlers/internal/storage/secrets_test.go:
```diff
@@ -734,6 +776,25 @@ func TestS3Extract_ForcePathStyle(t *testing.T) {
+			secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Data: map[string][]byte{
+					"endpoint":          []byte("https://vpce-1234567abc.s3.us-east-1.vpce.amazonaws.com"),
```
```suggestion
					"endpoint":          []byte("https://bucket.vpce-1234567abc.s3.us-east-1.vpce.amazonaws.com"),
```

JoaoBraveCoding reviewed (commented):
Thank @puretension for being such a great contributor, your words are much appreciated! 🙏 

Regarding support for the other endpoints unless there is a good technical reason I think we can postpone support for them. Mainly because this would also mean less code.

JoaoBraveCoding reviewed (dismissed):
Thank you so much for the quick fixes and being such an awesome contributor! From my side it's a lgtm but I'll wait for @xperimental to also give it a review.

JoaoBraveCoding commented:
Seems like we still have a failing test but apart from that should be good

puretension committed:
Fix VPC endpoint regex and test assertions

- Correct VPC endpoint regex to match actual AWS format (vpce-id vs bucket.vpce-id)
- Update test error messages to match implementation
- Fix TestS3Extract/aws_s3_vpc_endpoint_wrong_region test case

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension commented:
Hi @JoaoBraveCoding! 

I've fixed the failing test issue you mentioned. 
The problem was with the VPC endpoint regex pattern and test assertions. I've corrected both the regex to match actual AWS VPC endpoint format and updated the test error messages accordingly. 

Please let me know if there are any remaining issues or failure points.

@xperimental Could you please review this PR when you have a chance? 

Thanks for your patience and support! 🙏

puretension committed:
fix(operator): resolve merge conflict and fix VPC endpoint regex

Signed-off-by: puretension <rlrlfhtm5@gmail.com>

puretension committed:
Merge branch 'main' into fix/support-vpc-s3-endpoint-19243

Signed-off-by: Dohyeong Lee <rlrlfhtm5@gmail.com>

xperimental committed:
Merge branch 'main' into fix/support-vpc-s3-endpoint-19243

xperimental committed:
Merge branch 'main' into fix/support-vpc-s3-endpoint-19243

ptux commented:
Hi @puretension @JoaoBraveCoding @xperimental,

I've created a clean PR #22395 based on this work to help move this forward. It squashes the changes into a single commit and fixes a test bug in `TestS3Extract_ForcePathStyle/aws_s3_vpc_endpoint` (the expected `Endpoint` value should not include the `https://` scheme since `extractS3ConfigSecret` stores `parsedURL.Host`).

All tests pass locally. Hope this helps get the fix merged sooner!

xperimental commented:
Superseded by #22395, which has been merged.
